core/async/src/time.rs

```diff
@@ -198,19 +224,40 @@ impl FakeClock {
 
     /// Cancel-safe.
     pub async fn sleep(&self, d: Duration) {
-        let mut watch = self.0.lock().unwrap().mono.subscribe();
-        let t = *watch.borrow() + d;
-        while *watch.borrow() < t {
-            watch.changed().await.unwrap();
+        if d <= Duration::ZERO {
+            return;
         }
+        let receiver = {
+            let mut inner = self.0.lock().unwrap();
+            let (sender, receiver) = tokio::sync::oneshot::channel();
+            let waiter = ClockWaiterInHeap { waker: sender, deadline: inner.now() + d };
+            inner.waiters.push(waiter);
+            receiver
+        };
+        receiver.await.unwrap();
```

Comment:
nit: looks like this can be implemented by calling `sleep_until` with `inner.now() + d`, just to reduce code duplication